Unhandled 'error' event after sending shard

### Package Versions


```
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ npm list storj
storj-bridge@0.10.1 /vagrant
└── storj@2.0.7
```

Replace the values below using the output from `node --version`.

```
v4.4.3
```

### Expected Behavior

I should be able to send a shard to the farmer.

### Actual Behavior

After starting a bridge and farmers on the vagrant VM with npm run develop, when I come to sending a shard it crashes and turns off everything (Back to the command prompt). It prints out the following log:

```
`{"level":"info","message":"[minion] got network reply to 9f3c76b7a17354e593097b46efac9b29","timestamp":"2016-08-03T18:02:10.731Z"}
{"level":"info","message":"[minion] sending message to amq.gen-Iq5GzHCKSM2qVfj4iIL3wg","timestamp":"2016-08-03T18:02:10.731Z"}
{"level":"info","message":"Sending message to amq.gen-Iq5GzHCKSM2qVfj4iIL3wg","timestamp":"2016-08-03T18:02:10.732Z"}
{"level":"info","message":"Attempting to process message with correlationId 9f3c76b7a17354e593097b46efac9b29","timestamp":"2016-08-03T18:02:10.735Z"}
{"level":"info","message":"Processing message with correlationId 9f3c76b7a17354e593097b46efac9b29","timestamp":"2016-08-03T18:02:10.736Z"}
{"level":"info","message":"data channel connection opened","timestamp":"2016-08-03T18:02:10.958Z"}
{"level":"warn","message":"[renterpool] killing minions","timestamp":"2016-08-03T18:02:10.988Z"}
{"level":"info","message":"[minion] got work on storj.work.renterpool de1e9c441e853296a2e474b8b738d530","timestamp":"2016-08-03T18:02:10.995Z"}
events.js:141
      throw er; // Unhandled 'error' event
      ^

Error: write after end
    at writeAfterEnd (/vagrant/node_modules/readable-stream/lib/_stream_writable.js:201:12)
    at PassThrough.Writable.write (/vagrant/node_modules/readable-stream/lib/_stream_writable.js:244:20)
    at WebSocket.<anonymous> (/vagrant/node_modules/storj/lib/datachannel/server.js:177:17)
    at emitTwo (events.js:87:13)
    at WebSocket.emit (events.js:172:7)
    at Receiver.ontext (/vagrant/node_modules/storj/node_modules/ws/lib/WebSocket.js:841:10)
    at /vagrant/node_modules/storj/node_modules/ws/lib/Receiver.js:536:18
    at Receiver.applyExtensions (/vagrant/node_modules/storj/node_modules/ws/lib/Receiver.js:371:5)
    at /vagrant/node_modules/storj/node_modules/ws/lib/Receiver.js:508:14
    at Receiver.flush (/vagrant/node_modules/storj/node_modules/ws/lib/Receiver.js:347:3)
    at Receiver.opcodes.1.finish (/vagrant/node_modules/storj/node_modules/ws/lib/Receiver.js:541:12)
    at Receiver.expectHandler (/vagrant/node_modules/storj/node_modules/ws/lib/Receiver.js:493:33)
    at Receiver.add (/vagrant/node_modules/storj/node_modules/ws/lib/Receiver.js:103:24)
    at Socket.realHandler (/vagrant/node_modules/storj/node_modules/ws/lib/WebSocket.js:825:20)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:169:7)

npm ERR! Linux 3.13.0-92-generic
npm ERR! argv "/home/vagrant/.nvm/versions/node/v4.4.3/bin/node" "/home/vagrant/.nvm/versions/node/v4.4.3/bin/npm" "run" "develop"
npm ERR! node v4.4.3
npm ERR! npm  v2.15.1
npm ERR! code ELIFECYCLE
npm ERR! storj-bridge@0.10.1 develop: `STORJ_ALLOW_LOOPBACK=1 NODE_ENV=develop node script/develop.js start`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the storj-bridge@0.10.1 develop script 'STORJ_ALLOW_LOOPBACK=1 NODE_ENV=develop node script/develop.js start'.
npm ERR! This is most likely a problem with the storj-bridge package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     STORJ_ALLOW_LOOPBACK=1 NODE_ENV=develop node script/develop.js start
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs storj-bridge
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!
npm ERR!     npm owner ls storj-bridge
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:`
```


### Steps to Reproduce

vagrant up
vagrant ssh
npm run develop
send a shard using the java client available here: https://github.com/NutterzUK/storj-java-bridge-client

--- a/lib/datachannel/server.js
+++ b/lib/datachannel/server.js
@@ -203,7 +203,15 @@ DataChannelServer.prototype._handleConsignStream = function(socket, token) {
       hasher.update(chunk);
       item.shard.write(chunk);
 
-      if (received >= shardsize) {
+      if (received > shardsize) {
+        passthrough.removeAllListeners().end();
+        return socket.close(
+          DataChannelErrors.FAILED_INTEGRITY,
+          'The data transferred exceeds the amount defined in the contract'
+        );
+      }
+
+      if (received === shardsize) {
         passthrough.end();
       }
     }).resume();

--- a/lib/network/index.js
+++ b/lib/network/index.js
@@ -229,6 +229,7 @@ Network.prototype.leave = function(callback) {
     }
 
     if (self.tunclient && self.tunclient.status === TunnelClient.OPEN) {
+      self.tunclient.removeAllListeners();
       self.tunclient.close();
     }
 
@@ -697,12 +698,12 @@ Network.prototype._establishTunnel = function(tunnels, callback) {
       ));
     }
 
-    this.tunclient = new TunnelClient(
+    self.tunclient = new TunnelClient(
       tunnel,
       'http://127.0.0.1:' + localAddress.port
     );
 
-    this.tunclient.on('open', function() {
+    self.tunclient.on('open', function() {
       self._logger.info('tunnel successfully established: %j', alias);
 
       self._contact.address = alias.address;
@@ -713,7 +714,7 @@ Network.prototype._establishTunnel = function(tunnels, callback) {
       ContactChecker().check(self._contact, function(err) {
         if (err) {
           self._logger.warn('tunnel test failed, establishing new tunnel');
-          return this.tunclient.close();
+          return self.tunclient.close();
         }
 
         self._listenForTunnelers();
@@ -721,22 +722,22 @@ Network.prototype._establishTunnel = function(tunnels, callback) {
       });
     });
 
-    this.tunclient.on('close', function onTunnelClosed() {
+    self.tunclient.on('close', function onTunnelClosed() {
       self._logger.warn('tunnel connection closed');
-      this.tunclient.removeAllListeners('error');
+      self.tunclient.removeAllListeners('error');
       self._establishTunnel(tunnels, callback);
     });
 
-    this.tunclient.on('error', function onTunnelError(err) {
+    self.tunclient.on('error', function onTunnelError(err) {
       self._logger.warn(
         'tunnel connection lost, reason: %s',
         err.message
       );
-      this.tunclient.removeAllListeners();
+      self.tunclient.removeAllListeners();
       self._establishTunnel(tunnels, callback);
     });
 
-    this.tunclient.open();
+    self.tunclient.open();
   });
 };
 

--- a/lib/network/index.js
+++ b/lib/network/index.js
@@ -228,6 +228,10 @@ Network.prototype.leave = function(callback) {
       return callback(err);
     }
 
+    if (self.tunclient && self.tunclient.status === TunnelClient.OPEN) {
+      self.tunclient.close();
+    }
+
     self._node.disconnect(callback);
   });
 };
@@ -693,21 +697,23 @@ Network.prototype._establishTunnel = function(tunnels, callback) {
       ));
     }
 
-    var local = 'http://127.0.0.1:' + localAddress.port;
-    var tunclient = new TunnelClient(tunnel, local);
-    var checker = new ContactChecker();
+    this.tunclient = new TunnelClient(
+      tunnel,
+      'http://127.0.0.1:' + localAddress.port
+    );
 
-    tunclient.on('open', function() {
+    this.tunclient.on('open', function() {
       self._logger.info('tunnel successfully established: %j', alias);
 
       self._contact.address = alias.address;
       self._contact.port = alias.port;
 
       self._logger.info('testing newly established tunnel: %j', alias);
-      checker.check(self._contact, function(err) {
+
+      ContactChecker().check(self._contact, function(err) {
         if (err) {
           self._logger.warn('tunnel test failed, establishing new tunnel');
-          return tunclient.close();
+          return this.tunclient.close();
         }
 
         self._listenForTunnelers();
@@ -715,22 +721,22 @@ Network.prototype._establishTunnel = function(tunnels, callback) {
       });
     });
 
-    tunclient.on('close', function onTunnelClosed() {
+    this.tunclient.on('close', function onTunnelClosed() {
       self._logger.warn('tunnel connection closed');
-      tunclient.removeAllListeners('error');
+      this.tunclient.removeAllListeners('error');
       self._establishTunnel(tunnels, callback);
     });
 
-    tunclient.on('error', function onTunnelError(err) {
+    this.tunclient.on('error', function onTunnelError(err) {
       self._logger.warn(
         'tunnel connection lost, reason: %s',
         err.message
       );
-      tunclient.removeAllListeners();
+      this.tunclient.removeAllListeners();
       self._establishTunnel(tunnels, callback);
     });
 
-    tunclient.open();
+    this.tunclient.open();
   });
 };
 

--- a/lib/network/index.js
+++ b/lib/network/index.js
@@ -223,14 +223,14 @@ Network.prototype._enterOverlay = function(callback) {
 Network.prototype.leave = function(callback) {
   var self = this;
 
+  if (self._tunclient && self._tunclient.status === TunnelClient.OPEN) {
+    self._tunclient.removeAllListeners();
+    self._tunclient.close();
+  }
+
   this._manager.close(function(err) {
     if (err) {
       return callback(err);
-    }
-
-    if (self.tunclient && self.tunclient.status === TunnelClient.OPEN) {
-      self.tunclient.removeAllListeners();
-      self.tunclient.close();
     }
 
     self._node.disconnect(callback);
@@ -698,12 +698,12 @@ Network.prototype._establishTunnel = function(tunnels, callback) {
       ));
     }
 
-    self.tunclient = new TunnelClient(
+    self._tunclient = new TunnelClient(
       tunnel,
       'http://127.0.0.1:' + localAddress.port
     );
 
-    self.tunclient.on('open', function() {
+    self._tunclient.on('open', function() {
       self._logger.info('tunnel successfully established: %j', alias);
 
       self._contact.address = alias.address;
@@ -714,7 +714,7 @@ Network.prototype._establishTunnel = function(tunnels, callback) {
       ContactChecker().check(self._contact, function(err) {
         if (err) {
           self._logger.warn('tunnel test failed, establishing new tunnel');
-          return self.tunclient.close();
+          return self._tunclient.close();
         }
 
         self._listenForTunnelers();
@@ -722,22 +722,22 @@ Network.prototype._establishTunnel = function(tunnels, callback) {
       });
     });
 
-    self.tunclient.on('close', function onTunnelClosed() {
+    self._tunclient.on('close', function onTunnelClosed() {
       self._logger.warn('tunnel connection closed');
-      self.tunclient.removeAllListeners('error');
+      self._tunclient.removeAllListeners('error');
       self._establishTunnel(tunnels, callback);
     });
 
-    self.tunclient.on('error', function onTunnelError(err) {
+    self._tunclient.on('error', function onTunnelError(err) {
       self._logger.warn(
         'tunnel connection lost, reason: %s',
         err.message
       );
-      self.tunclient.removeAllListeners();
+      self._tunclient.removeAllListeners();
       self._establishTunnel(tunnels, callback);
     });
 
-    self.tunclient.open();
+    self._tunclient.open();
   });
 };
 

--- a/lib/network/transport.js
+++ b/lib/network/transport.js
@@ -8,6 +8,7 @@ var ip = require('ip');
 var TunnelServer = require('../tunnel/server');
 var merge = require('merge');
 var utils = require('../utils');
+var ContactChecker = require('./contactchecker');
 
 /**
  * Custom HTTP transport adapter
@@ -58,34 +59,67 @@ inherits(Transport, kad.transports.HTTP);
 Transport.prototype._open = function(callback) {
   var self = this;
 
-  if (ip.isPublic(self._contact.address) || self._noforward) {
+  if (self._noforward) {
     self._isPublic = true;
 
-    /* istanbul ignore else */
-    if (ip.isPrivate(self._contact.address)) {
-      self._log.warn(
-        'you are not bound to a public address, traversal strategies disabled'
-      );
-    }
+    /* istanbul ignore next */
+    self._log.warn(
+      'your address is %s and traversal strategies are disabled',
+      ip.isPublic(self._contact.address) ? 'public' : 'private'
+    );
 
     return kad.transports.HTTP.prototype._open.call(self, callback);
   }
 
   self._requiresTraversal = true;
 
-  self._log.warn(
-    'you are not bound to a public address, trying traversal strategies...'
-  );
-  self._forwardPort(function(err, wanip, port) {
-    self._isPublic = !err;
+  function _traverseNat() {
+    self._log.warn(
+      'you are not publicly reachable, trying traversal strategies...'
+    );
+    self._forwardPort(function(err, wanip, port) {
+      self._isPublic = !err;
 
-    if (self._isPublic) {
-      self._contact.port = port || self._contact.port;
-      self._log.info('kad node bound and port mapped: %s', self._contact.port);
+      if (self._isPublic) {
+        self._contact.port = port || self._contact.port;
+        self._log.info('node bound and port mapped: %s', self._contact.port);
+      }
+
+      kad.transports.HTTP.prototype._open.call(self, callback);
+      self._contact.address = wanip || self._contact.address;
+    });
+  }
+
+  kad.transports.HTTP.prototype._open.call(self, function() {
+    self._checkIfReachable(function(isReachable) {
+      if (isReachable) {
+        return callback(null);
+      }
+
+      kad.transports.HTTP.prototype._close.call(self);
+      _traverseNat();
+    });
+  });
+};
+
+/**
+ * Checks if we are publicly reachable
+ * @private
+ * @param {Function}
+ */
+Transport.prototype._checkIfReachable = function(callback) {
+  if (ip.isPrivate(this._contact.address)) {
+    return callback(false);
+  }
+
+  var checker = new ContactChecker();
+
+  checker.check(this._contact, function(err) {
+    if (err) {
+      return callback(false);
     }
 
-    kad.transports.HTTP.prototype._open.call(self, callback);
-    self._contact.address = wanip || self._contact.address;
+    callback(true);
   });
 };
 

--- a/lib/tunnel/client.js
+++ b/lib/tunnel/client.js
@@ -31,9 +31,13 @@ function TunnelClient(tunnel, target) {
   this._muxer = null;
   this._demuxer = null;
   this._channels = {};
+  this.status = TunnelClient.CLOSED;
 
   events.EventEmitter.call(this);
 }
+
+TunnelClient.OPEN = 1;
+TunnelClient.CLOSED = 0;
 
 /**
  * Triggered when the tunnel is opened
@@ -92,6 +96,7 @@ TunnelClient.prototype.open = function() {
   });
 
   this._tunnel.on('open', function() {
+    this.status = TunnelClient.OPEN;
     self.emit('open');
   });
 
@@ -118,7 +123,8 @@ TunnelClient.prototype.close = function() {
   }
 
   this._tunnel = null;
-
+  this.status = TunnelClient.CLOSED;
+  
   this.emit('close');
 };
 

--- a/package.json
+++ b/package.json
@@ -86,7 +86,7 @@
     "semver": "^5.1.0",
     "through": "^2.3.8",
     "tmp": "0.0.28",
-    "ws": "^1.0.1"
+    "ws": "^1.1.0"
   },
   "devDependencies": {
     "browserify": "^11.1.0",

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "storj",
-  "version": "2.1.0",
+  "version": "2.1.1",
   "description": "implementation of the storj protocol for node.js and the browser",
   "main": "index.js",
   "directories": {

--- a/test/keyring.unit.js
+++ b/test/keyring.unit.js
@@ -20,6 +20,7 @@ var tmpfolder = function() {
 };
 
 describe('KeyRing', function() {
+  this.timeout(6000);
 
   after(function() {
     cleanup.forEach(function(folder){

--- a/test/network/index.unit.js
+++ b/test/network/index.unit.js
@@ -124,6 +124,26 @@ describe('Network (public)', function() {
       expect(net._node.disconnect.called).to.equal(true);
     });
 
+    it('should close the tunnel client if it is open', function() {
+      var net = Network({
+        keypair: KeyPair(),
+        manager: Manager(RAMStorageAdapter()),
+        logger: kad.Logger(0),
+        seeds: [],
+        address: '127.0.0.1',
+        port: 0,
+        noforward: true
+      });
+      net._tunclient = {
+        status: 1,
+        close: sinon.stub(),
+        removeAllListeners: sinon.stub()
+      };
+      net._node = { disconnect: sinon.stub() };
+      net.leave();
+      expect(net._tunclient.close.called).to.equal(true);
+    });
+
     it('should callback with error if db close fails', function(done) {
       var net = Network({
         keypair: KeyPair(),

--- a/test/network/transport.unit.js
+++ b/test/network/transport.unit.js
@@ -31,14 +31,92 @@ describe('Network/Transport', function() {
         Transport.prototype,
         '_forwardPort'
       ).callsArg(0);
+      var _checkIfReachable = sinon.stub(
+        Transport.prototype,
+        '_checkIfReachable'
+      ).callsArgWith(0, false);
       var transport = new Transport(Contact({
         address: '127.0.0.1',
         port: 0,
         nodeID: KeyPair().getNodeID()
       }));
       transport.on('ready', function() {
-        expect(_forwardPort.called).to.equal(true);
         _forwardPort.restore();
+        _checkIfReachable.restore();
+        expect(_forwardPort.called).to.equal(true);
+        done();
+      });
+    });
+
+    it('should not close the transport if already reachable', function(done) {
+      var _checkIfReachable = sinon.stub(
+        Transport.prototype,
+        '_checkIfReachable'
+      ).callsArgWith(0, true);
+      var _close = sinon.stub(
+        require('kad').transports.HTTP.prototype,
+        '_close'
+      );
+      var transport = new Transport(Contact({
+        address: '127.0.0.1',
+        port: 0,
+        nodeID: KeyPair().getNodeID()
+      }));
+      transport.on('ready', function() {
+        _close.restore();
+        _checkIfReachable.restore();
+        expect(_close.called).to.equal(false);
+        done();
+      });
+    });
+
+  });
+
+  describe('#_checkIfReachable', function() {
+
+    it('it should check the contact if address public', function(done) {
+      var _check = sinon.stub().callsArg(1);
+      var StubbedTransport = proxyquire('../../lib/network/transport', {
+        ip: { isPrivate: sinon.stub().returns(false) },
+        './contactchecker': sinon.stub().returns({
+          check: _check
+        })
+      });
+      StubbedTransport.prototype._checkIfReachable.call({
+        _contact: { address: 'public.ip.address' }
+      }, function() {
+        done();
+      });
+    });
+
+    it('it should callback true if reachable', function(done) {
+      var _check = sinon.stub().callsArgWith(1, null);
+      var StubbedTransport = proxyquire('../../lib/network/transport', {
+        ip: { isPrivate: sinon.stub().returns(false) },
+        './contactchecker': sinon.stub().returns({
+          check: _check
+        })
+      });
+      StubbedTransport.prototype._checkIfReachable.call({
+        _contact: { address: 'public.ip.address' }
+      }, function(result) {
+        expect(result).to.equal(true);
+        done();
+      });
+    });
+
+    it('it should callback false if reachable', function(done) {
+      var _check = sinon.stub().callsArgWith(1, new Error('Failed'));
+      var StubbedTransport = proxyquire('../../lib/network/transport', {
+        ip: { isPrivate: sinon.stub().returns(false) },
+        './contactchecker': sinon.stub().returns({
+          check: _check
+        })
+      });
+      StubbedTransport.prototype._checkIfReachable.call({
+        _contact: { address: 'public.ip.address' }
+      }, function(result) {
+        expect(result).to.equal(false);
         done();
       });
     });

